src/discrete_alloc.rs

```diff
@@ -0,0 +1,271 @@
+use std::alloc::{self, Layout};
+use std::sync;
+
+use nix::sys::mman::ProtFlags;
+
+use crate::helpers::ToU64;
+
+static ALLOCATOR: sync::Mutex<MachineAlloc> = sync::Mutex::new(MachineAlloc::empty());
+
+/// A distinct allocator for the `MiriMachine`, allowing us to manage its
+/// memory separately from that of Miri itself.
+#[derive(Debug)]
+pub struct MachineAlloc {
+    pages: Vec<*mut u8>,
+    huge_allocs: Vec<(*mut u8, usize)>,
+    allocated: Vec<Box<[u8]>>,
+    page_size: usize,
+    enabled: bool,
+}
+
+// SAFETY: We only point to heap-allocated data
+unsafe impl Send for MachineAlloc {}
+
+impl MachineAlloc {
+    // Allocation-related methods
+
+    /// Initializes the allocator with placeholder 4k pages.
+    const fn empty() -> Self {
+        Self {
+            pages: Vec::new(),
+            huge_allocs: Vec::new(),
+            allocated: Vec::new(),
+            page_size: 4096,
+            enabled: false,
+        }
+    }
+
+    /// SAFETY: There must be no existing `MiriAllocBytes`
+    pub unsafe fn enable() {
+        let mut alloc = ALLOCATOR.lock().unwrap();
+        alloc.enabled = true;
+        // This needs to specifically be the system pagesize!
+        alloc.page_size = unsafe {
+            let ret = libc::sysconf(libc::_SC_PAGE_SIZE);
+            if ret > 0 {
+                ret.try_into().unwrap()
+            } else {
+                4096 // fallback
```

Comment:
Given that this apparently must be the page size, how is this fallback justified?

Reply:
The easiest thing would be to just panic instead. Or are there real situations where we cannot determine the page size?

Reply:
I guess I was just paranoid of extra unwraps, but you're probably right. I'll change this